### PR TITLE
Allow cache dirs outside the main source dir

### PR DESCRIPTION
**Pull request by jeremy**
Better for e.g. Guard and Watchman since they needn't get gummed up watching large deep cache dirs.

TODO: Setting the cache dir is untested currently, so we'll need to add coverage before demonstrating that the feature behaves as expected.

### CODE REVIEW

**Comment by envygeeks** [comment]
If you would still like this merge can you please rebase so that I can merge it? Thanks!

**Comment by jeremy** [comment]
Rebased 👍

**Comment by jeremy** [comment]
@envygeeks Rebased 👍

**Comment by envygeeks** [comment]
@jeremy looks like these tests might have failed, I've restarted them just in-case, if they fail again mind taking a look at what might have broken?

**Comment by jeremy** [comment]
@envygeeks Fixed it up.

Note the compat issue here. Other dirs are given relative to the root dir, but cache was configurable relative to the source dir only.

To preserve compat, we check whether the cache dir exists relative to the source dir first, then resolve it relative to the root dir if that doesn't exist. So existing apps use the same cache and fresh apps use a cache in the root dir.

File: lib/jekyll/assets/env.rb [hunk 1]
```diff
@@ -142,9 +148,9 @@ def liquid_proxies
       # @return [Pathname/Pathutil]
       # --
       def in_cache_dir(*paths)
-        jekyll.in_source_dir(asset_config["cache"] || ".asset-cache",
-          *paths
-        )
+        paths.reduce(@cache_path) do |base, path|
+          Jekyll.sanitized_path(base, path)
+        end
```
Comment (jeremy) [inline lib/jekyll/assets/env.rb]: Mimics `Jekyll::Site#in_source_dir` implementation. Resolves path relative to the cache path, which is no longer required to live within the source dir.

File: lib/jekyll/assets/env.rb [hunk 2]
```diff
@@ -111,6 +111,12 @@ def initialize(path, jekyll = nil)
         path ? super(path) : super()
         @jekyll = jekyll
 
+        # Analagous to Jekyll::Site @source and @destination paths
+        @cache_path = asset_config.fetch("cache", ".asset-cache")
+        if File.exist?(cache_path_in_source_dir = jekyll.in_source_dir(@cache_path))
+          @cache_path = cache_path_in_source_dir
+        end
+
```
Comment (jeremy) [inline lib/jekyll/assets/env.rb]: For back compat, resolve the cache dir against the source path first. Resolve the cache relative to the root dir if an old source-relative cache path doesn't already exist.

File: .gitignore [hunk 3]
```diff
@@ -1,4 +1,5 @@
 /.analysis
+/.asset-cache
```
Comment (jeremy) [inline .gitignore]: Shush, toplevel cache path.

File: lib/jekyll/assets/liquid/tag/proxied_asset.rb [hunk 4]
```diff
@@ -155,7 +155,7 @@ def find_cached
           def cache_file
             return true if @_cached || find_cached
 
-            filename.dirname.mkdir_p unless filename.dirname == env.in_cache_dir
+            filename.dirname.mkdir_p
```
Comment (jeremy) [inline lib/jekyll/assets/liquid/tag/proxied_asset.rb]: Don't require the cache dir to already exist. Create it on demand.

**Comment by envygeeks** [comment]
I didn't get to this today, but I'm updating to let you know I didn't forget.

**Comment by envygeeks** [comment]
Thank :heart:

**Pull request merged**